### PR TITLE
OTT-703 :: Dynamic price floor implementation

### PR DESCRIPTION
**Pull request by pm-nikhil-vaidya**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: floors/PriorityFetchHeap.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package floors
+
+import "github.com/prebid/prebid-server/config"
```
Comment (PubMatic-OpenWrap) [inline floors/PriorityFetchHeap.go:3]: rename file to priority_heap.go

File: floors/PriorityFetchHeap.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	FetchPeriod int64
+}
+
+type FetchQueue []*FetchInfo
+
+func (fq FetchQueue) Len() int {
+	return len(fq)
+}
+
+func (fq FetchQueue) Less(i, j int) bool {
```
Comment (PubMatic-OpenWrap) [inline floors/PriorityFetchHeap.go:16]: Make private functions, as these function will be used only in floors package

File: floors/PriorityFetchHeap.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import "github.com/prebid/prebid-server/config"
+
+type FetchInfo struct {
+	config.AccountFloorFetch
+	FetchPeriod int64
+}
+
+type FetchQueue []*FetchInfo
+
+func (fq FetchQueue) Len() int {
```
Comment (pm-jaydeep-mohite) [inline floors/PriorityFetchHeap.go:12]: Make private functions, as these function will be used only in floors package

File: floors/fetcher.go [hunk 4]
```diff
@@ -1,135 +1,213 @@
...
-		return []error{fmt.Errorf("Empty data in floors JSON  in JSON Unmarshall ")}
+	if priceFloors.Data == nil {
+		return errors.New("empty data in floor JSON")
 	}
 
-	var validModelGroups []openrtb_ext.PriceFloorModelGroup
-	for _, modelGroup := range floorData.ModelGroups {
-		if len(modelGroup.Values) > fetchConfig.MaxRules {
-			err = append(err, fmt.Errorf("Number of rules = %v in modelgroup are greater than limit = %v", len(modelGroup.Values), fetchConfig.MaxRules))
+	if len(priceFloors.Data.ModelGroups) < 1 {
```
Comment (pm-jaydeep-mohite) [inline floors/fetcher.go:188]: Can be updated to 
if  len(priceFloors.Data.ModelGroups) == 0

File: floors/fetcher.go [hunk 5]
```diff
@@ -1,135 +1,213 @@
...
+		return errors.New("no model groups found in price floor data")
+	}
+
+	for _, modelGroup := range priceFloors.Data.ModelGroups {
+		if len(modelGroup.Values) < 1 || len(modelGroup.Values) > configs.MaxRules {
+			return errors.New("invalid number of floor rules, floor rules should be greater than zero and less than MaxRules specified in account config")
+		}
+
+		if modelGroup.ModelWeight != nil {
+			if *modelGroup.ModelWeight < 1 || *modelGroup.ModelWeight > 100 {
```
Comment (pm-jaydeep-mohite) [inline floors/fetcher.go:198]: above if check can be clubbed with this if

File: floors/fetcher.go [hunk 6]
```diff
@@ -1,135 +1,213 @@
...
-	fetchStatus string                      `json:"fetchstatus,omitempty"`
+type FloorFetcher interface {
+	Fetch(configs config.AccountPriceFloors) *openrtb_ext.PriceFloorRules
 }
 
-var fetchInProgress map[string]bool
-
-func fetchInit() {
-	fetchInProgress = make(map[string]bool)
+type PriceFloorFetcher struct {
```
Comment (Pubmatic-Dhruv-Sonone) [inline floors/fetcher.go:23]: Add comments on struct fields

File: floors/fetcher.go [hunk 7]
```diff
@@ -1,135 +1,213 @@
...
 
-	if !validator.IsURL(fetchConfig.URL) {
-		return &fetchResult{
-			fetchStatus: openrtb_ext.FetchError,
-		}
+func (f *PriceFloorFetcher) Fetch(configs config.AccountPriceFloors) *openrtb_ext.PriceFloorRules {
+
+	//check in cache: hit/miss
+	//hit: directly return
+	// pwd, _ := os.Getwd()
```
Comment (Pubmatic-Dhruv-Sonone) [inline floors/fetcher.go:48]: Remove comments code.

File: floors/fetcher.go [hunk 8]
```diff
@@ -1,135 +1,213 @@
...
 
-	timeout := (time.Duration(account.PriceFloors.Fetch.Timeout) * time.Millisecond)
-	if timeout > 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithDeadline(ctx, start.Add(timeout))
-		defer cancel()
+func (f *PriceFloorFetcher) priceFloorFetcher() {
+
+	//Create Ticker of 5 seconds
+	ticker := time.NewTicker(5 * time.Second)
```
Comment (Pubmatic-Dhruv-Sonone) [inline floors/fetcher.go:102]: Make ticket interval constant

File: floors/PriorityFetchHeap.go (hunk 3 above)
Comment (pm-nikhil-vaidya) [reply floors/PriorityFetchHeap.go:12]: Here we are using container/heap package
All these methods are being called by heap implementation
That's why they are exposed

File: floors/fetcher.go [hunk 9]
```diff
@@ -1,135 +1,213 @@
...
-	if !fetchInprogress {
-		fetchPriceFloorRulesAsynchronous(account)
-	}
+	return nil
+}
 
-	// Rules not present in cache, fetch rules asynchronously
-	return &fetchResult{
-		fetchStatus: openrtb_ext.FetchInprogress,
+func (f *PriceFloorFetcher) floorfetcherWorker(configs config.AccountFloorFetch) {
```
Comment (pm-nikhil-vaidya) [inline floors/fetcher.go:72]: rename to worker

File: floors/fetcher.go [hunk 10]
```diff
@@ -1,135 +1,213 @@
...
+					heap.Push(&f.fetchQueue, &nextFetch)
+				}
+			}
+		case <-f.done:
+			glog.Info("Price Floor fetcher terminated")
+		}
 	}
+}
+
+func floorFetcherAndValidator(configs config.AccountFloorFetch) *openrtb_ext.PriceFloorRules {
```
Comment (pm-nikhil-vaidya) [inline floors/fetcher.go:126]: Rename to fetchAndValidate

File: main.go [hunk 11]
```diff
@@ -54,7 +55,9 @@ func serve(cfg *config.Configuration) error {
 	currencyConverterTickerTask := task.NewTickerTask(fetchingInterval, currencyConverter)
 	currencyConverterTickerTask.Start()
 
-	_, err := router.New(cfg, currencyConverter)
+	priceFloorFetcher := floors.NewPriceFloorFetcher(20, 1000)
```
Comment (pm-nikhil-vaidya) [inline main.go:58]: Initialise inside router

Comment (pm-nikhil-vaidya) [inline main.go:58]: if it starts to fail, then server should not start

File: floors/fetcher.go [hunk 12]
```diff
@@ -1,135 +1,213 @@
...
+	}
 
-		if len(respBody) > account.PriceFloors.Fetch.MaxFileSize {
-			return []error{fmt.Errorf("Receieved more than MaxFileSize")}
-		}
+	return &priceFloors
+}
 
-		var priceFloors openrtb_ext.PriceFloorRules
+func fetchPriceFloorRulesFromURL(URL string, timeout int) ([]byte, error) {
```
Comment (pm-jaydeep-mohite) [inline floors/fetcher.go:154]: remove PriceFloors from all function name

File: floors/fetcher.go (hunk 4 above)
Comment (pm-nikhil-vaidya) [reply floors/fetcher.go]: Done

File: floors/fetcher.go (hunk 5 above)
Comment (pm-nikhil-vaidya) [reply floors/fetcher.go]: done

File: floors/fetcher.go (hunk 6 above)
Comment (pm-nikhil-vaidya) [reply floors/fetcher.go:23]: done

File: floors/fetcher.go (hunk 8 above)
Comment (pm-nikhil-vaidya) [reply floors/fetcher.go:105]: Will decide the time interval and make it constant

File: main.go (hunk 11 above)
Comment (pm-nikhil-vaidya) [reply main.go]: DONE

File: floors/fetcher.go (hunk 12 above)
Comment (pm-nikhil-vaidya) [reply floors/fetcher.go]: DONE

File: floors/PriorityFetchHeap.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
+package floors
+
+import "github.com/prebid/prebid-server/config"
+
+type FetchInfo struct {
+	config.AccountFloorFetch
```
Comment (pm-jaydeep-mohite) [inline floors/PriorityFetchHeap.go:6]: can we move code into fetcher.go  ?

**Pull request closed without merging**